clippy_lints/src/methods/mod.rs

```diff
@@ -1428,6 +1429,23 @@ fn lint_unnecessary_fold(cx: &LateContext<'_, '_>, expr: &hir::Expr, fold_args:
         return;
     }
 
+    // `Iterator::any` cannot be used within a pattern guard
+    // See https://github.com/rust-lang-nursery/rust-clippy/issues/3069
+    if_chain! {
+        if let Some(fold_parent) = cx.tcx.hir.find(cx.tcx.hir.get_parent_node(expr.id));
```

Comment:
This probably doesn't work if the pattern guard is even a slight variation of the test (so e.g. `((0..3).fold(false, |acc, x| acc || x > 2), 42).0`. I don't see a good way to fix this issue. It seems like an unfortunate restriction of the language that will get fixed at some point. I think users should rather disable the lint than us trying to hide the false positive.

It also seems like there's ongoing work in the compiler towards fixing this: https://play.rust-lang.org/?gist=1ec006ebd816427e4acbf21f2156d6e2&version=nightly&mode=debug&edition=2015

We could list this as a known issue in the lint description.

Reply:
I see; that is indeed an issue. If there is an upcoming compiler feature to fix this, then I agree, I would rather wait for that, and update the documentation instead.